File: PPMEditor.py
#!/usr/bin/env python3
import sys
import math

# Kendall Nicley
# COSC 365 Prog. Languages
# Lab 6: PPM Editor using python
# This lab is to take input to alter a PPM image
# The alterations are mirror along x and y axis
# and invert the color of the image


#Pixel class the will contian the values of the pixels
#and will return the color values of the pixel
class Pixel:
	def __init__(self, red, green, blue):
		self.codic = {"r":red,"g":green,"b":blue}

#Accesors and Mutators that will get and set the pixel color value
	def setRed(self, arg): #Red pixel
		self.codic["r"] = arg
	def getRed(self):
		return self.codic["r"]
	def setGreen(self, arg): #Green Pixel
		self.codic["g"] = arg
	def getGreen(self):
		return self.codic["g"]
	def setBlue(self, arg): # Blue Pixel
		self.codic["b"] = arg
	def getBlue(self):
		return self.codic["b"]

#The PPM class that will hold the functions
class PPM:
	#Default overloaded constructor that will assign the image width
	# and height and create an empty list for the pixels
	def __init__(self, w, h, mi):
		self.width = w
		self.height = h
		self.maxI = mi
		self.pixlist= []
	
	#This function is to invert the color of each pixel
	# Within the pixel list.
	def invert(self):
		#Get the value of the maximum intensity to use in the inversion
		max = self.getMaxIntensity()

		#Loop through the pixels and subtract from the Maximum intesity to get
		# Inverted value
		for y in range(0, self.getHeight()):
			for x in range(0, self.getWidth()):
				pix = self.getPixel(x,y)

				#get RGB values
				r = pix.getRed()
				g = pix.getGreen()
				b = pix.getBlue()

				# Invert the values in each color
				pix.setRed(max - r)
				pix.setGreen(max - g)
				pix.setBlue(max - b)
		
	#This function will mirror the given image along the y axis
	#It will loop halfway throught the rows and swap the pixels left over right
	def flipHorizontal(self):
		#Loop through the columns
		for y in range(0, self.getHeight()):
			#loop through the rows
			for x in range(0, int(math.floor(self.getWidth()/2))):
				# Set the pixel to tmp
				tmp = self.getPixel(x,y)
				# swap the pixels
				self.setPixel(x,y,(self.getPixel((x*-1)-1, y)))
				self.setPixel((x*-1)-1,y,(tmp))



	#Function will mirror along the x axis
	#It will loop halfway through the height and swap the top and bottom half
	def flipVertical(self):
		mid = int(math.floor(self.getHeight()/2));
		# for each of the columns
		for y in range(0, int(math.floor(self.getHeight()/2))):
			#loop throught each of the rows
			for x in range(0, self.getWidth()):
				#swap the pixels
				tmp = self.getPixel(x,y)
				self.setPixel(x,y,(self.getPixel(x, (y * -1)-1)));
				self.setPixel(x,(y * -1)-1,(tmp));
				
	#Add pixel to the internal list
	def addPixel(self, Pixel):
		self.pixlist.append(Pixel)

	#Remove the pixels from the list
	def clearPixels(self):
		self.pixlist.clear()

	#Return the pixel from the internal list
	def getPixel(self, x, y):
		#Do a bound test and throw error if there is a problem
		if x > self.width or y > self.height:
			raise IndexOutOfRangeException

		# Calculate the index
		index = x + y * self.width
		try:
			return self.pixlist[index]
		except ValueError:
			print(ValueError)
	
	#Set the pixel to the given value
	def setPixel(self, x,y,Pixel):
		# bound check of the rows and columns throw error if over
		if x > self.width or y > self.height:
			raise IndexOutOfRangeException
		else:
			#calculate index for setting
			index = x + y * self.width
			try:
				self.pixlist[index] = Pixel
			except ValueError:
				print(ValueError)


	#Return the height of the image
	def getHeight(self):
		return self.height

	#Return the width of the image
	def getWidth(self):
		return self.width

	#Return the Maximum Intensity of the image
	def getMaxIntensity(self):
		return self.maxI

#This function will read the given file and create the values
def readFile(inFile):
	file = open(inFile, "r")
	file.readline().strip()
	
	#Read the Width and Height
	val = file.readline().strip().split()
	wid = int(val[0])
	height = int(val[1])

	#Read the maximum intensity
	val = file.readline().strip()
	maxi = int(val)

	#Create the ppm editor and pass with width, height, and max intensity
	ppm = PPM(wid, height, maxi) 

	#For each line in the file
	for line in file:
		val = line.strip().split() #Split each line to list elements(not encluding \n)
		red = int(val[0]) #Assign the red pixel
		green = int(val[1]) #Assign the green pixel
		blue = int(val[2]) #Assign the blue pixel
		ppm.addPixel(Pixel(red,green,blue))

	return ppm
		
#Write the new PPM File
def writePPM(fileName,ppm):

	#Open and write to the given output file
	file = open(fileName, "w")
	file.write("P3\n")
	file.write(str(ppm.getWidth()) + " " + str(ppm.getHeight()) + "\n")
	file.write(str(ppm.getMaxIntensity()) + "\n")

	
	#Scan through each columns
	for y in range(0, ppm.getHeight()):
		#Scan through each row
		for x in range(0, ppm.getWidth()):
			#write each of the pixels to the file
			#print(x + ppm.getWidth() * y)
			pix = ppm.getPixel(x,y)
			file.write(str(pix.getRed()) + " " + str(pix.getGreen()) + " " + str(pix.getBlue()) + "\n")
	
#create the PPM Image
def createPPM(argv):

	#Bound check the input
	if (len(argv) < 4):
		print("Error: Not enough arguments")
		print("  ./PPMEditor inputFile outputFile {Flag(V,FH,FV,or N)}")
		print("  Exiting!")
		return 0;

	#assign the given inputs
	inputF = argv[1]
	outputF = argv[2]
	ppm = readFile(inputF)
	
	#Determine which flag was given and do approiate action
	if (argv[3] == "V"):
		ppm.invert()
	elif (argv[3] == "FH"):
		ppm.flipHorizontal()
	elif (argv[3] == "FV"):
		ppm.flipVertical()
	elif (argv[3] == "N"):
		ppm = ppm
	
	#Call the writer to the file
	writePPM(outputF, ppm)	
	
#Main function
def main():
	createPPM(sys.argv)
	return 0

#main()
if __name__ == "__main__":
	exit(main())

File: PPMEditor.js
#!/usr/bin/env nodejs
//Kendall Nicley
//Cosc 365 Lab 5 PPM Editor
//This lab is to make a image editor
//for .ppm files and will be able
//to mirror along x and y axis and invert
//the colors

'use strict';

const fs = require('fs');

//Main function
function main(argc, argv)
{
	//calling process
	processPPM(argc,argv);


	return argc;
}

//Process is the entry point for the PPM processor and will
//manage the given flags and create the required components
function processPPM(argc, argv)
{
	if (argv.length > 4)
	{
		var inputF = argv[2];
		var outputF = argv[3];
		var flag = argv[4];
		

		//Read the file	
		var data = readPPM(inputF, outputF, flag);

	}else
		console.log("Error to Few arguments given. Exiting!");

	return 0;
}

//Write the new PPM to the file
function writePPM(ppm, outputF)
{
	//Create the stream to write the file
	var fout = fs.createWriteStream(outputF, {flags: 'w', encoding: 'utf8', fd: null, mode:0o644});

	//Close the file
	fout.on('close', () => {console.log("File Closed");});

	fout.write(`P3\n${ppm.getWidth()} ${ppm.getHeight()}\n${ppm.getMaxIntensity()}\n`,'utf8');

	var numRows = ppm.getHeight();
	var numCols = ppm.getWidth();
	//Read through each of the rows and columns
	for (var row = 0; row < numRows; row++)
	{
		for (var col = 0; col < numCols; col++)
		{
			var pix = ppm.getPixel(row,col);
			//console.log(pix.getRed() + " " + pix.getGreen() + " " + pix.getBlue());
			fout.write(`${pix.getRed()} ${pix.getGreen()} ${pix.getBlue()}\n`);
		}
	}
}


//Function to create the PPM from the given file
function readPPM(inputF, outputF, flag)
{
	//String to be passed
	var ret = "";

	//Create the stream reader for the file
	var stream = fs.createReadStream(inputF, {flags: 'r', encoding: 'utf8'});	

	//Error check the file
	stream.on('error', function() {console.log(`Error: ${inputF} Does not exists. Exiting!`)})
	//Holder for ppm values
	stream.on('data', (chunk) => 
			{
				//Convert chunk to string and add to buffer
				ret = ret.concat(chunk.toString('utf8'));
			}); //End of Fin.on Data function


	stream.on('close', () =>
			{
				var ppm;
				//This is before the timer implementation
				//ppm = processFile(ret, outputF, flag);	
				//writePPM(ppm, outputF);

				//Create a timer for processing file
				var id = setImmediate(function()
				{
					ppm = processFile(ret, outputF, flag);	
					
				});
				
				//Wait a second to allow the setImmediate to work
				setTimeout(() => {if(ppm !== null)
									writePPM(ppm, outputF)}, 1);
			});
			

}

function processFile(buffer, outFile, flag)
{
	//Replace new line with spaces and break into elements seperated by spaces
	var data = buffer.replace(new RegExp("\n", 'g')," ").split(" ");

	//Read The header and Creat the PPM
	var width = data[1];
	var height = data[2];
	var MaxI = data[3];
	var ppm = new PPM(width, height, MaxI);

	
	
	//Read through the data and assign the pixel
	for(var i = 4; i < data.length - 1; i+= 3)
	{
		//Read each pixel in the file
		var pix = new Pixel(data[i], data[i+1], data[i+2]);
		ppm.addPixel(pix); //Add to the internal array
	}

	//Read the flags
	if (flag === "V") //Invert the given ppm file
		ppm.invert()
	else if (flag === "FH") //Flip around the y axis
		ppm.flipHorizontal();
	else if (flag === "FV") //Flip around the X axis
		ppm.flipVertical();
	else if(flag !== "N") //If it's N for nothing, then it's an error
	{
		console.log(`Error: Unidentified Flag: ${flag}\n Exiting!`);
		return null;
	}

	return ppm;

}

//Pixel class that contains the Pixel properties
//and accessors and mutators
class Pixel
{
	constructor(r=0,g=0,b=0,)
	{
		this.redChan = r;
		this.greenChan = g;
		this.blueChan = b;
	}

	//Mutators and accessors for the pixel colors
	getRed()  { return this.redChan; };
	setRed(i)  { this.redChan = i; };
	getGreen() { return this.greenChan; };
	setGreen(i) { this.greenChan = i; };
	getBlue() { return this.blueChan; };
	setBlue(i) { this.blueChan = i; };
}

class PPM
{
	//array of pixels and values height, width, 
	constructor(w,h,mi)
	{
		this.Height = h;
		this.Width = w;
		this.MaxIntensity = mi;
		this.pixlist =[];
	}

	//This function will subtract the color from
	//the maximum Intensity to get the oppisite color
	invert()
	{
		var max = this.getMaxIntensity();

		//Read through each of the rows and columns
		var numRows = this.getHeight();
		var numCols = this.getWidth();
		var pix;
		for (var row = 0; row < numRows; row++)
		{
			for (var col = 0; col < numCols; col++)
			{
				//Get the pixel and set the color
				pix = this.getPixel(row,col);
				pix.setRed(max-pix.getRed());
				pix.setGreen(max-pix.getGreen());
				pix.setBlue(max-pix.getBlue());
			}
		}
	}

	//This function will scan each row and swap the
	//pixels around the center(left to right) and mirror
	//the image horizontally
	flipHorizontal()
	{
		//get the size of image
		var numRows = this.getHeight();
		var numCols = this.getWidth();
		for (var row = 0; row < numRows; row++) //Scan the rows
		{
			for (var col = 0; col < numCols/2; col++) //Scan the columns
			{
				//Get the pixels to swap and swap them
				var pix = this.getPixel(row, col);
				var tmp = this.getPixel(row, numCols - (col+1));
				this.setPixel(row,col, tmp);
				this.setPixel(row,numCols - (col + 1), pix);
			}
		}

	}
	
	//This function will scan each col and swap each pixels
	//around the center(top to bottom) and mirror the image
	//vertically
	flipVertical()
	{
		//get the size of image
		var numRows = this.getHeight();
		var numCols = this.getWidth();
		for (var row = 0; row < numRows/2; row++) //Scan the rows
		{
			for (var col = 0; col < numCols; col++) //Scan the columns
			{
				//Get and set the pixels to be swapped
				var pix = this.getPixel(row, col);
				var tmp = this.getPixel(numRows - (row+1), col);
				this.setPixel(row,col, tmp);
				this.setPixel(numRows - (row + 1), col, pix);
			}
		}
	}
	
	//Add Pixel to the pixel list
	addPixel(pix) { this.pixlist.push(pix);}

	//Delete the array of pixels
	clearPixels() { this.pixlist=[]; }

	//Return the pixel at the x,y value
	getPixel(x,y)
	{
		//Calculate the index
		var index = x * this.Width + y;

		//Check the index to make sure it's valid
		if (index > this.pixlist.length)
		{throw({name:"IndexOutOfRangeException", message:`Index was out of range of the pixel array: X:${x} Y:${y} Index:${index} length:${this.pixlist.length}`})}
		else
			return this.pixlist[index];
	}

	//Set the pixel at the x,y and assign the given pixel class
	setPixel(x,y,pix)
	{
		//Calculate the index
		var index = x * this.Width + y;
		
		//Check if index is within the approiate range
		if (index > this.pixlist.length)
		{throw({name:"IndexOutOfRangeException", message:"Index was out of range of the pixel array"})}
		else
			this.pixlist[index] = pix;

	}

	getHeight() { return this.Height; } //Height of the image
	getWidth() { return this.Width; } //The width of the image
	getMaxIntensity() { return this.MaxIntensity; } //The maximum Intensity
}

//Call the function 
const ret = main(process.argv.length, process.argv);
